Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,600 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2019 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import NIO
+import NIOConcurrencyHelpers
+import NIOHTTP1
+import NIOTLS
+
+/// A connection pool that manages and creates new connections to hosts respecting the specified preferences
+class ConnectionPool {
+    /// The configuration used to bootstrap new HTTP connections
+    private let configuration: HTTPClient.Configuration
+
+    /// The main data structure used by the `ConnectionPool` to retreive and create connections associated
+    /// to a a given `Key` .
+    /// - Warning: This property should be accessed with proper synchronization, see `connectionProvidersLock`
+    var connectionProviders: [Key: HTTP1ConnectionProvider] = [:]
```

Comment:
this property should be made `private` because it must be accessed with the `connectionProvidersLock` which is `private`. So there's no correct use of it outside this `class` anyway.

Reply:
Tried it and it's actually used by a test, so I wasn't properly syncing it in the test (even if in practice there was no issue because it was timed correctly), but I think I'll make the lock internal then

Reply:
that works or you make it private and export a

```swift
var testsOnly_connectionProviders: ... = {
    return self.lock.withLock {
        return self.connectionProviders
    }
}
```